src/Forms/FieldList.php

```diff
@@ -77,18 +120,67 @@ public function dataFields()
      */
     public function saveableFields()
     {
-        if (!$this->sequentialSaveableSet) {
-            $this->collateDataFields($this->sequentialSaveableSet, true);
+        if (empty($this->sequentialSaveableSet)) {
+            $fields = [];
+            $this->recursiveWalk(function (FormField $field) use (&$fields) {
+                if (!$field->canSubmitValue()) {
+                    return;
+                }
+                $name = $field->getName();
+                if (isset($fields[$name])) {
+                    $this->fieldNameError($field, __FUNCTION__);
+                }
+                $fields[$name] = $field;
+            });
+            $this->sequentialSaveableSet = $fields;
         }
         return $this->sequentialSaveableSet;
     }
 
+    /**
+     * Return array of all field names
+     *
+     * @return array
+     */
+    public function fieldNames()
+    {
+        return array_keys($this->dataFields());
+    }
+
+    protected function fieldNameError(FormField $field, $functionName)
+    {
+        if ($this->form) {
+            $errorSuffix = sprintf(
+                " in your '%s' form called '%s'",
+                get_class($this->form),
+                $this->form->getName()
+            );
+        } else {
+            $errorSuffix = '';
+        }
+
+        user_error(
+            sprintf(
```

Comment:
Should we use exceptions, so that user-code can catch this?

Reply:
Possibly, what exception? `DuplicateFieldNameException`?

Reply:
is that a breaking change? maybe we can't do that...

Reply:
Well, this error breaks the code, so it's going to break anyway.

Reply:
`Exception` is fine, or even `LogicException`. You don't need a separate class.